discovery_child_development/notebooks/02_explore_concepts.py

```diff
@@ -11,164 +13,302 @@
+# We will check Kmeans results as well in case this method performs better than HDBSCAN.
+
+# %%
+# Define K-Means search parameters
+kmeans_search_params = {
+    "n_clusters": [5, 10, 15, 20, 25, 30, 35],
+    "init": ["k-means++"],
+}
+
+# %%
+# %%time
+# Parameter grid search using K-Means
+kmeans_search_results = kmeans_param_grid_search(
+    vectors=unique_concepts_50, search_params=kmeans_search_params, random_seeds=[SEED]
 )
-unique_concepts_df["color"] = unique_concepts_df["cluster"].apply(
-    lambda x: "noise" if x == -1 else x
+
+# %%
+# The silhouette scores for HDBSCAN turn out to be better
+kmeans_search_results
+
+# %%
+# proceed with HDBSCAN clustering, seeing as that leads to a better silhouette score
+hdbscan_labels = cau.hdbscan_clustering(
+    unique_concepts_50, hdbscan_params=optimal_hdbscan_params, have_noise_labels=False
 )
 
-# -
+# %%
+hdbscan_labels.head()
+
+# %%
+# Reduce original vectors to 2D for plotting
+unique_concepts_2d = cau.reduce_to_2D(unique_concepts_384, random_state=SEED)
 
-# In the plot below:
-# * cluster 28 covers computing-related things
-# * cluster 20 is statistics! These concepts probably come up in any paper that has a vagiely quant angle
-# * 5 is cohort and longitudinal studies (very relevant when studying children's development)
-# * Artificial intelligence ends up in the noise cluster
+# %%
+# Add 2D vectors into the dataframe for the sake of plotting. Add probability to the dataframe as well for later use
+unique_concepts_df = unique_concepts_df.assign(
+    cluster=hdbscan_labels["labels"],
+    probability=hdbscan_labels["probability"],
+    x_hdbscan=unique_concepts_2d[:, 0],
+    y_hdbscan=unique_concepts_2d[:, 1],
+)
 
-# +
-fig1 = (
-    alt.Chart(unique_concepts_df)
+# %%
+fig_hdbscan = (
+    alt.Chart(
+        unique_concepts_df[unique_concepts_df["probability"] > 0.1]
+    )  # filtering out some rows because altair has a limit on how many data points you can plot
     .mark_circle()
     .encode(
-        x="x",
-        y="y",
-        color=alt.Color("color:N", legend=alt.Legend(title="cluster")),
-        opacity="opacity:Q",
-        tooltip=["concept", "cluster"],
+        x="x_hdbscan",
+        y="y_hdbscan",
+        color=alt.Color("cluster:N", legend=alt.Legend(title="cluster")),
+        tooltip=["display_name", "cluster"],
     )
     .properties(width=800, height=600)
     .interactive()
 )
 
-fig1
-# -
+fig_hdbscan
+
+# %% [markdown]
+# # Check which concepts were assigned to which cluster
+#
+# In this section, we check the probability with which concepts were assigned to their clusters. It turns out there is a bimodal distribution of probabilities (probably because we forced everything to be assigned to a cluster instead of allowing a noise cluster?). We divide concepts into "core" and "peripheral", where "core" ones are ones that were assigned to their clusters with a high probability, and "peripheral" are the ones that were assigned with a low probability.
```

Comment:
This is what ChatGPT thinks about the bimodal distribution - I guess, it could be a somewhat discrete division between whether a point happened to be part of the cluster's dendrogram segment or not? (depending on the dendrogram depth). So, it's not based on metric distance (which would create an expectation of a smoothly distributed probabilities) but more of discrete clustering outcome. I'm not 100% clear, but perhaps helpful for intuition and further reading... 

> ## Soft Clustering and Probabilities:
> HDBSCAN has a feature known as "soft clustering", which provides for each data point a probability of belonging to each cluster. This is unlike "hard clustering" where each point is definitively assigned to a single cluster.
> 
> The soft clustering feature works by considering the hierarchy of the dendrogram. The idea is simple: if a data point was part of a cluster early in the hierarchy (i.e., when it was more stringent), then it's more likely to truly belong to that cluster. However, if it only became part of a cluster later in the hierarchy (when it was more lenient), then its membership is more uncertain.
> 
> ## Bimodal Probability Distribution:
> When you observe a bimodal distribution with lots of probabilities close to 0 or 1, it suggests that:
> 
> Points with Probabilities Close to 1: These are the core samples or points that are well within the density of a particular cluster. They have been part of their respective clusters from early on in the hierarchy and thus have a high probability of truly belonging to those clusters.
> 
> Points with Probabilities Close to 0: These points either belong to the noise or are on the boundaries between clusters. They were either not part of any cluster for a significant portion of the hierarchy, or they frequently shifted between different clusters as the hierarchy was built. Thus, their true cluster membership is uncertain, reflected by their low probabilities.

Reply:
Thanks Karlis, that information is useful. It sounds like HDBSCAN proceeds iteratively and the earlier the iteration when a datapoint was assigned to a cluster, the higher the probability?

It'd be interesting to try this on some different data to see if hdbscan probabilities always look like this or whether it was something particular to this data. A lot of the concepts contain the same or very similar words so I can imagine that HDBSCAN would have assigned them to clusters almost immediately.